File: flutter/lib/services/image_processor.dart
import 'dart:typed_data';
import 'package:image/image.dart' as img;

/// Resultado de procesar una foto: bytes ya redimensionados/comprimidos,
/// listos para guardar localmente y subir.
class ImagenProcesada {
  final Uint8List bytes;
  final String mimeType;
  ImagenProcesada(this.bytes, this.mimeType);
}

enum TipoFoto { cedulaFrontal, cedulaPosterior, fotoParticipante }

/// Redimensiona + comprimve una foto ANTES de guardarla localmente o
/// subirla. Todo en Dart puro (paquete `image`), funciona igual en
/// Android/iOS/Web -- a diferencia de flutter_image_compress, que en Web
/// tiene soporte limitado.
///
/// Estrategia: apuntar a 300-800 KB via calidad JPEG decreciente, pero
/// nunca sacrificar legibilidad por debajo de una calidad minima -- si con
/// la calidad minima sigue pesando mas, se deja mas grande (ver pedido del
/// usuario: "si una imagen necesita mas tamaño para mantener la
/// informacion legible, permite un tamaño mayor").
class ImageProcessor {
  static const int _objetivoMinBytes = 300 * 1024;
  static const int _objetivoMaxBytes = 800 * 1024;
  static const int _calidadMinima = 55;

  static Future<ImagenProcesada> procesarCedula(Uint8List original) =>
      _procesar(original, ladoMayorMax: 1600);

  static Future<ImagenProcesada> procesarPerfil(Uint8List original) =>
      _procesar(original, ladoMayorMax: 1200);

  static Future<ImagenProcesada> _procesar(
    Uint8List original, {
    required int ladoMayorMax,
  }) async {
    final decodificada = img.decodeImage(original);
    if (decodificada == null) {
      throw const FormatException('No se pudo leer la imagen capturada.');
    }

    // decodeImage ya aplica la orientacion EXIF; al reencodear como JPEG
    // sin copiar el EXIF original, se descartan metadatos (GPS, modelo de
    // camara, etc.) que no hacen falta para este uso.
    final ladoMayor = decodificada.width > decodificada.height
        ? decodificada.width
        : decodificada.height;
    final redimensionada = ladoMayor > ladoMayorMax
        ? img.copyResize(
            decodificada,
            width: decodificada.width >= decodificada.height ? ladoMayorMax : null,
            height: decodificada.height > decodificada.width ? ladoMayorMax : null,
            interpolation: img.Interpolation.average,
          )
        : decodificada;

    // Busqueda simple de calidad: empieza alto y baja hasta entrar en el
    // rango objetivo o tocar el piso de calidad legible.
    Uint8List mejorIntento = Uint8List(0);
    for (int calidad = 90; calidad >= _calidadMinima; calidad -= 5) {
      final codificada = Uint8List.fromList(
        img.encodeJpg(redimensionada, quality: calidad),
      );
      mejorIntento = codificada;
      if (codificada.lengthInBytes <= _objetivoMaxBytes) break;
    }

    // Si incluso a calidad minima sigue por debajo del piso esperable
    // (imagen muy simple/chica), no hace falta forzar mas compresion.
    if (mejorIntento.lengthInBytes < _objetivoMinBytes &&
        mejorIntento.lengthInBytes == 0) {
      mejorIntento = Uint8List.fromList(img.encodeJpg(redimensionada, quality: 85));
    }

    return ImagenProcesada(mejorIntento, 'image/jpeg');
  }

  static String nombreArchivo(TipoFoto tipo) {
    switch (tipo) {
      case TipoFoto.cedulaFrontal:
        return 'cedula_frontal.jpg';
      case TipoFoto.cedulaPosterior:
        return 'cedula_posterior.jpg';
      case TipoFoto.fotoParticipante:
        return 'perfil.jpg';
    }
  }

  static String claveTipo(TipoFoto tipo) {
    switch (tipo) {
      case TipoFoto.cedulaFrontal:
        return 'cedula_frontal';
      case TipoFoto.cedulaPosterior:
        return 'cedula_posterior';
      case TipoFoto.fotoParticipante:
        return 'foto_participante';
    }
  }
}

File: src/lib/image-processor.ts
// Redimensiona + comprime una foto en el navegador (canvas) antes de
// guardarla/subirla -- espejo de flutter/lib/services/image_processor.dart.
// Apunta a 300-800 KB via calidad JPEG decreciente, sin bajar de una
// calidad minima legible (si con esa calidad sigue pesando mas, se deja
// mas grande en vez de sacrificar legibilidad).

const OBJETIVO_MAX_BYTES = 800 * 1024;
const CALIDAD_MINIMA = 0.55;

export type TipoFoto = "cedula_frontal" | "cedula_posterior" | "foto_participante";

export function nombreArchivo(tipo: TipoFoto): string {
  return `${tipo}.jpg`;
}

async function cargarImagen(archivo: Blob): Promise<HTMLImageElement> {
  const url = URL.createObjectURL(archivo);
  try {
    const img = new Image();
    img.src = url;
    await img.decode();
    return img;
  } finally {
    URL.revokeObjectURL(url);
  }
}

function canvasABlob(canvas: HTMLCanvasElement, calidad: number): Promise<Blob> {
  return new Promise((resolve, reject) => {
    canvas.toBlob(
      (blob) => (blob ? resolve(blob) : reject(new Error("No se pudo codificar la imagen."))),
      "image/jpeg",
      calidad
    );
  });
}

async function procesar(archivo: Blob, ladoMayorMax: number): Promise<Blob> {
  const img = await cargarImagen(archivo);
  const ladoMayor = Math.max(img.width, img.height);
  const escala = ladoMayor > ladoMayorMax ? ladoMayorMax / ladoMayor : 1;

  const canvas = document.createElement("canvas");
  canvas.width = Math.round(img.width * escala);
  canvas.height = Math.round(img.height * escala);
  const ctx = canvas.getContext("2d");
  if (!ctx) throw new Error("No se pudo procesar la imagen (canvas no disponible).");
  // dibujar en un canvas nuevo y reencodear como JPEG descarta metadatos
  // (EXIF/GPS) que no hacen falta para este uso.
  ctx.drawImage(img, 0, 0, canvas.width, canvas.height);

  let mejorIntento: Blob | null = null;
  for (let calidad = 0.9; calidad >= CALIDAD_MINIMA; calidad -= 0.05) {
    const blob = await canvasABlob(canvas, calidad);
    mejorIntento = blob;
    if (blob.size <= OBJETIVO_MAX_BYTES) break;
  }
  return mejorIntento!;
}

export async function procesarCedula(archivo: Blob): Promise<Blob> {
  return procesar(archivo, 1600);
}

export async function procesarPerfil(archivo: Blob): Promise<Blob> {
  return procesar(archivo, 1200);
}
